### DISCUSSION TITLE
Component re-render when non-selected store store state change.

### DISCUSSION THREAD

**Discussion by Fawwaz-2009** [pmndrs/zustand#1728] [Q&A] [answered]
Hi all 👋 , I'm trying to understand why a component is re-rendering even thought the selected state has not changed!. it was my understand that re-rendering will only be triggered for changes in the selected state and not on every store state change. 

To illustrate the point I made a very short example where we have two component a toggle (update state on click) and another named ShouldRenderOnce which only use the part of the store that should not change. However if you see the console, you would notice that anytime the toggle is clicked both components re-render!. here is the link to the sandbox as well as some code snippets, https://codesandbox.io/s/zustand-react-demo-forked-v5kzou?file=/src/components/ShouldRenderOnce.js

![image](https://user-images.githubusercontent.com/32019791/229323978-6dea249a-2e5e-4cb1-8503-c00ac86259f8.png)

![image](https://user-images.githubusercontent.com/32019791/229323983-1229b86a-6fc7-4fd9-b5af-26eaf05d600c.png)

![image](https://user-images.githubusercontent.com/32019791/229323987-0afeefa7-41fd-4918-94e2-63148f32b2e5.png)

**Comment by JacobWeisenburger** [comment, marked as answer]
Fixed: https://codesandbox.io/s/zustand-react-demo-forked-up3qvi?file=/src/components/ToggleTheme.js

the problem was that you had `const darkTheme = configStore((state) => state.darkTheme);` in `App` and so that was rerendering all of `App`.

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by Fawwaz-2009** [reply]
Well that was a stupid mistake 😄 . Thanks @JacobWeisenburger .

So the initial reason why I started looking into this was because I didn't understand how zustand is able to isolate re-renders like this?. I'm going off the custom hook mental model and following that if the hook re-render the component using the hook will re-render. That doesn't appear to be the case in Zustand. **To illustrate the issue I wrapped the Zustand store state in a custom hook and keeping pretty much the same selector api, and now the re-renders are not isolated to selected properties!**.

![image](https://user-images.githubusercontent.com/32019791/229671411-640dc99e-6b3a-44e7-9a6f-6e5f7c0ee8e7.png)



I'm aware that Zustand uses useSyncExternalStore, but from the react docs it doesn't appear that useSyncExternalStore should have any more features above what included in useState or useReducer  [Note Section from the docs](https://react.dev/reference/react/useSyncExternalStore#subscribing-to-an-external-store)

Here is the link for the new repo: https://codesandbox.io/s/zustand-react-demo-forked-rb0ogb?file=/src/stores/configStore.js


PS: thanks again for your help and no worries if you don't know the answer, just let me know will open a new question for this.

**Reply by JacobWeisenburger** [reply]
I think I know the answer, but please open a new question. I'll answer there.

**Reply by Fawwaz-2009** [reply]
done https://github.com/pmndrs/zustand/discussions/1732
